File: Sprint6/D/C++/main.cpp
// https://contest.yandex.ru/contest/25069/problems/D/
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

void bfs(int start_v, const vector<vector<int>> &g) {
    int n = g.size() - 1;
    vector<bool> visited(n + 1, false);
    queue<int> q;

    q.push(start_v);
    visited[start_v] = true;

    while (!q.empty()) {
        int current_v = q.front();
        q.pop();

        cout << current_v << " ";

        for (int neighbor : g[current_v]) {
            if (!visited[neighbor]) {
                q.push(neighbor);
                visited[neighbor] = true;
            }
        }
    }

    cout << endl;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> g(n + 1);

    // Чтение рёбер
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    for (int i = 1; i <= n; ++i) {
        sort(g[i].begin(), g[i].end());
    }

    int start_v;
    cin >> start_v;

    bfs(start_v, g);

    return 0;
}

File: Sprint6/D/Rust/src/main.rs
// https://contest.yandex.ru/contest/25069/problems/D/

use std::collections::VecDeque;
use std::io::{self, BufRead};

fn bfs(start_v: usize, g: &[Vec<usize>]) {
    let mut visited = vec![false; g.len()];
    let mut q = VecDeque::new();

    q.push_back(start_v);
    visited[start_v] = true;

    while let Some(current_v) = q.pop_front() {
        print!("{} ", current_v + 1);

        for &neighbor in &g[current_v] {
            if !visited[neighbor] {
                q.push_back(neighbor);
                visited[neighbor] = true;
            }
        }
    }

    println!();
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let first_line = lines.next().unwrap().unwrap();
    let nm: Vec<usize> = first_line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let (n, m) = (nm[0], nm[1]);

    let mut g: Vec<Vec<usize>> = vec![vec![]; n];

    for _ in 0..m {
        let edge: Vec<usize> = lines
            .next()
            .unwrap()
            .unwrap()
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap() - 1)
            .collect();
        let (u, v) = (edge[0], edge[1]);
        g[u].push(v);
        g[v].push(u);
    }

    for neighbors in g.iter_mut() {
        neighbors.sort();
    }

    let start_v: usize = lines.next().unwrap().unwrap().parse::<usize>().unwrap() - 1;
    bfs(start_v, &g);
}
